File: StudioProject/SimpleGithubApp/app/src/main/java/skim/dev/kr/simplegithubapp/ui/repo/RepositoryActivity.java
package skim.dev.kr.simplegithubapp.ui.repo;

import android.app.Activity;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.View;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import skim.dev.kr.simplegithubapp.R;
import skim.dev.kr.simplegithubapp.api.GithubApi;
import skim.dev.kr.simplegithubapp.api.GithubApiProvider;
import skim.dev.kr.simplegithubapp.api.model.GithubRepo;
import skim.dev.kr.simplegithubapp.ui.GlideApp;

/**
 * Created by suyoung on 2018. 2. 26..
 */

public class RepositoryActivity extends Activity {

    public static final String KEY_USER_LOGIN = "SwimingKim";
    public static final String KEY_RERO_NAME = "Android-Java";

    LinearLayout llContent;
    ImageView ivProfile;
    TextView tvName;
    TextView tvStars;
    TextView tvDescription;
    TextView tvLanguage;
    TextView tvLastUpdate;
    ProgressBar pbProgress;
    TextView tvMessage;
    GithubApi api;
    Call<GithubRepo> repoCall;

    SimpleDateFormat dateFormatInResponse = new SimpleDateFormat("yyyy-MM-dd HH:mm:ssX", Locale.getDefault());
    SimpleDateFormat dateFormatToShow = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault());

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_repository);

        llContent = (LinearLayout)findViewById(R.id.llActivityRepositoryContent);
        ivProfile = (ImageView)findViewById(R.id.ivActivityRepositoryProfile);
        tvName = (TextView)findViewById(R.id.tvActivityRepositoryName);
        tvStars = (TextView)findViewById(R.id.tvActivityRepositoryStars);
        tvDescription = (TextView)findViewById(R.id.tvActivityRepositoryDescription);
        tvLanguage = (TextView)findViewById(R.id.tvActivityRepositoryLanguage);
        tvLastUpdate = (TextView)findViewById(R.id.tvActivityRepositoryLastUpdate);
        pbProgress = (ProgressBar)findViewById(R.id.pbActivityRepository);
        tvMessage = (TextView)findViewById(R.id.tvActivityRepositoryMessage);
        api = GithubApiProvider.provideGithubApi(this);

        String login = getIntent().getStringExtra(KEY_USER_LOGIN);
        if (null == login) {
            throw new IllegalArgumentException("No login into exists in extras");
        }
        String repo = getIntent().getStringExtra(KEY_RERO_NAME);
        if (null == repo) {
            throw new IllegalArgumentException("No repo info exists in extras");
        }

        showRepositoryInfo(login, repo);
    }

    private void showRepositoryInfo(String login, final String repoName) {
        showProgress();

        repoCall = api.getRepository(login, repoName);
        repoCall.enqueue(new Callback<GithubRepo>() {
            @Override
            public void onResponse(Call<GithubRepo> call, Response<GithubRepo> response) {
                hideProgress(true);

                GithubRepo repo = response.body();
                if (response.isSuccessful() && null != repo) {
                    GlideApp.with(RepositoryActivity.this).load(repo.owner.avatarUrl).into(ivProfile);

                    tvName.setText(repo.fullName);
                    tvStars.setText(getResources().getQuantityString(R.plurals.star, repo.stars, repo.stars));
                    if (null == repo.description) {
                        tvDescription.setText(R.string.no_description_provided);
                    } else {
                        tvDescription.setText(repo.description);
                    }
                    if (null == repo.language) {
                        tvLanguage.setText(R.string.no_language_specified);
                    } else {
                        tvLanguage.setText(repo.language);
                    }

                    try {
                        Date lastUpdate = dateFormatInResponse.parse(repo.updatedAt);
                        tvLastUpdate.setText(dateFormatToShow.format(lastUpdate));
                    } catch (ParseException e) {
                        tvLastUpdate.setText(getString(R.string.unknown));
                    }
                } else {
                    showError("Not succesful : "+response.message());
                }
            }

            @Override
            public void onFailure(Call<GithubRepo> call, Throwable t) {
                hideProgress(false);
                showError(t.getMessage());
            }
        });


    }

    private void showProgress() {
        llContent.setVisibility(View.GONE);
        pbProgress.setVisibility(View.VISIBLE);
    }

    private void hideProgress(boolean isSucced) {
        llContent.setVisibility(isSucced ? View.VISIBLE : View.GONE);
        pbProgress.setVisibility(View.GONE);
    }

    private void showError(String message) {
        tvMessage.setText(message);
        tvMessage.setVisibility(View.VISIBLE);
    }
}

File: StudioProject/SimpleGithubApp/app/src/main/kotlin/skim/dev/kr/simplegithubapp/ui/repo/RepositoryActivity.kt
package skim.dev.kr.simplegithubapp.ui.repo

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.View
import io.reactivex.android.schedulers.AndroidSchedulers
import kotlinx.android.synthetic.main.activity_repository.*
import skim.dev.kr.simplegithubapp.R
import skim.dev.kr.simplegithubapp.api.provideGithubApi
import skim.dev.kr.simplegithubapp.rx.AutoClearedDisposable
import skim.dev.kr.simplegithubapp.extensions.plusAssign
import skim.dev.kr.simplegithubapp.ui.GlideApp
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*

/**
 * Created by suyoung on 2018. 2. 26..
 */

class RepositoryActivity : AppCompatActivity() {

    companion object {

        const val KEY_USER_LOGIN = "user_login"
        const val KEY_RERO_NAME = "repo_name"
    }

    internal val api by lazy { provideGithubApi(this) }
    internal val disposables = AutoClearedDisposable(this)

    internal val dateFormatInResponse = SimpleDateFormat("yyyy-MM-dd HH:mm:ssX", Locale.getDefault())
    internal val dateFormatToShow = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault())

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_repository)

        lifecycle += disposables

        val login = intent.getStringExtra(KEY_USER_LOGIN)
                ?: throw IllegalArgumentException("No login into exists in extras")
        val repo = intent.getStringExtra(KEY_RERO_NAME)
                ?: throw IllegalArgumentException("No repo info exists in extras")

        showRepositoryInfo(login, repo)
    }

    private fun showRepositoryInfo(login: String, repoName: String) {

        disposables += api.getRepository(login, repoName)
                .observeOn(AndroidSchedulers.mainThread())
                .doOnSubscribe { showProgress() }
                .doOnError { hideProgress(false) }
                .doOnComplete { hideProgress(true) }
                .subscribe({ repo ->
                    GlideApp.with(this@RepositoryActivity)
                            .load(repo.owner.avatarUrl)
                            .into(ivActivityRepositoryProfile)

                    tvActivityRepositoryName.text = repo.fullName
                    tvActivityRepositoryStars.text = resources.getQuantityString(R.plurals.star, repo.stars, repo.stars)
                    if (null == repo.description) {
                        tvActivityRepositoryDescription.setText(R.string.no_description_provided)
                    } else {
                        tvActivityRepositoryDescription.text = repo.description
                    }
                    if (null == repo.language) {
                        tvActivityRepositoryLanguage.setText(R.string.no_language_specified)
                    } else {
                        tvActivityRepositoryLanguage.text = repo.language
                    }

                    try {
                        val lastUpdate = dateFormatInResponse.parse(repo.updatedAt)
                        tvActivityRepositoryLastUpdate.text = dateFormatToShow.format(lastUpdate)
                    } catch (e: ParseException) {
                        tvActivityRepositoryLastUpdate.text = getString(R.string.unknown)
                    }
                }) {
                    showError(it.message)
                }


    }

    private fun showProgress() {
        llActivityRepositoryContent.visibility = View.GONE
        pbActivityRepository.visibility = View.VISIBLE
    }

    private fun hideProgress(isSucced: Boolean) {
        llActivityRepositoryContent.visibility = if (isSucced) View.VISIBLE else View.GONE
        pbActivityRepository.visibility = View.GONE
    }

    private fun showError(message: String?) {

        with(tvActivityRepositoryMessage) {
            text = message ?: "Unexpected error."
            visibility = View.VISIBLE
        }
    }

}
